### ISSUE TITLE
Contains on a readonly collection throws InvalidOperation exception.

### ISSUE THREAD

**Issue by leniency** [dotnet/efcore#18658]
A query that uses a `collection.Contains` will fail when the collection is marked `readonly`. It works as expected without the `readonly`.

### Steps to reproduce

```csharp
public class QueryTests
{
    // If this is marked "readonly", the query will fail. Remove "readonly" and it works.
    static readonly IEnumerable<TagCategoryType> Types =
        new TagCategoryType[] { TagCategoryType.Custom, TagCategoryType.Subject };

    [TestMethod]
    public void EFCoreQuery_WhereIn()
    {
        using var db = GetContext();
        var tags = db.Set<Tag>()
            .Where(t => Types.Contains(t.Type))
            .ToList();

        Assert.IsNotNull(tags);
        Assert.IsTrue(tags.Count > 0);
    }
}
```

#### Exception
```
Message: 
    Test method Enroller.Tests.Dev.EntityFrameworkCore.QueryTests.EFCoreQuery_WhereIn threw exception: 
    System.InvalidOperationException: Processing of the LINQ expression 'AsQueryable<TagCategoryType>((IEnumerable<TagCategoryType>)TagCategoryType[] { Custom, SecurityGroup, })' by 'NavigationExpandingExpressionVisitor' failed. This may indicate either a bug or a limitation in EF Core. See https://go.microsoft.com/fwlink/?linkid=2101433 for more detailed information.
  Stack Trace: 
    NavigationExpandingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
    MethodCallExpression.Accept(ExpressionVisitor visitor)
    ExpressionVisitor.Visit(Expression node)
    NavigationExpandingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
    MethodCallExpression.Accept(ExpressionVisitor visitor)
    ExpressionVisitor.Visit(Expression node)
    NavigationExpandingExpressionVisitor.ExpandNavigationsInExpression(NavigationExpansionExpression source, Expression expression)
    NavigationExpandingExpressionVisitor.ExpandNavigationsInLambdaExpression(NavigationExpansionExpression source, LambdaExpression lambdaExpression)
    NavigationExpandingExpressionVisitor.ProcessWhere(NavigationExpansionExpression source, LambdaExpression predicate)
    NavigationExpandingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
    MethodCallExpression.Accept(ExpressionVisitor visitor)
    ExpressionVisitor.Visit(Expression node)
    NavigationExpandingExpressionVisitor.ExpandAndReduce(Expression query, Boolean applyInclude)
    NavigationExpandingExpressionVisitor.Expand(Expression query)
    QueryTranslationPreprocessor.Process(Expression query)
    QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
    Database.CompileQuery[TResult](Expression query, Boolean async)
    QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
    <>c__DisplayClass9_0`1.<Execute>b__0()
    CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
    CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
    QueryCompiler.Execute[TResult](Expression query)
    EntityQueryProvider.Execute[TResult](Expression expression)
    EntityQueryable`1.GetEnumerator()
    List`1.ctor(IEnumerable`1 collection)
    Enumerable.ToList[TSource](IEnumerable`1 source)
   ```

### Further technical details

EF Core version: 3.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework:  .NET Core 3.0
Operating system: Windows 10
IDE: Visual Studio 2019 16.3

**Comment by smitpatel** [comment]
readonly here factors as inline-ing the whole expression rather than creating the parameter. The reason we fail to translate is, there is an additional cast to `IEnumerable<TagCategoryType>` in front of the array.

@leniency - As a work-around (if you cannot remove readonly), you can also change the type of `Types` to be `TagCategoryType[]`.

**Comment by AbrahamLopez10** [comment]
Hi @smitpatel , I also encountered this issue. I'm running .NET Core 3.0 and EF Core 3.0.

Basically, code like the one below fails saying the LINQ expression cannot be converted to SQL and that I need to use `.AsEnumerable()`:

```cs
private ICollection<App> GetAppsDictionary(ICollection<string> appNames)
{
    return _dbContext.Set<App>.Where(x => appNames.Contains(x.Name))
        .ToDictionary(x => x.Name, x => x);
}
```

In the code above, if I change `appNames` from `ICollection<string>` to just `string[]` then the error is gone and the LINQ expression is translated to SQL.

I'd have imagined that EF Core would be smart enough to translate `ICollection` just like any other enumerable? This works correctly in EF 6 but for some reason doesn't work in EF Core?

**Comment by wdnijdam** [comment]
See #19320.
A IReadOnlyCollection as local variable is accepted but a (static) field throws an error.

**Comment by ajcvickers** [comment]
See also the scenario in #23123.

**Comment by smitpatel** [comment]
This may be fixed in #25577
